### PR TITLE
Add timestamp tracking

### PR DESCRIPTION
**Pull request by ansd**

### CODE REVIEW

**Comment by michaelklishin** [comment]
@ansd can you please rebase this on top of `master`? 🙏

**Comment by ansd** [comment]
I rebased and force-pushed.

File: src/osiris.erl [hunk 1]
```diff
@@ -133,13 +134,14 @@ start_replica(Replica, Config) ->
 write(Pid, WriterId, Corr, Data) ->
     osiris_writer:write(Pid, self(), WriterId, Corr, Data).
 
--spec write_tracking(pid(), binary(), offset()) -> ok.
-write_tracking(Pid, TrackingId, Offset) ->
-    osiris_writer:write_tracking(Pid, TrackingId, Offset).
+-spec write_tracking(pid(), binary(), {tracking_type(), offset() | timestamp()}) -> ok.
+write_tracking(Pid, TrackingId, {_TrkType, _TrkData} = Tracking) ->
```
Comment (kjnilsson) [inline src/osiris.erl:138]: This will break code that uses it in rabbit

File: src/osiris.erl [hunk 2]
```diff
@@ -133,13 +134,14 @@ start_replica(Replica, Config) ->
...
+    osiris_writer:write_tracking(Pid, TrackingId, Tracking).
 
--spec read_tracking(pid(), binary()) -> {tracking_type(), offset()} | undefined.
-read_tracking(Pid, TrackingId) ->
-    osiris_writer:read_tracking(Pid, TrackingId).
+-spec read_tracking(pid(), tracking_type(), binary()) ->
+    {tracking_type(), offset() | timestamp()} | undefined.
+read_tracking(Pid, TrackingType, TrackingId) ->
```
Comment (kjnilsson) [inline src/osiris.erl:143]: This will also break code in rabbit. Perhaps we should retain backwards compat  by handling the previous call expectations.

**Review by kjnilsson — CHANGES_REQUESTED** [review]
I think it may make sense to retain backwards compatibility.

File: src/osiris.erl (hunk 1 above)
Comment (ansd) [reply src/osiris.erl:139]: Pushed a commit to retain backwards compatibility.

File: src/osiris.erl (hunk 2 above)
Comment (ansd) [reply src/osiris.erl:156]: Pushed a commit to retain backwards compatibility.

**Comment by ansd** [comment]
Thank you for reviewing 🙂
In that case we could either close or merge https://github.com/rabbitmq/rabbitmq-server/pull/3217.
Given that Osiris itself is `v1.0.0` as I just learnt, it's good practice to retain backwards compatibility 👍

**Review by kjnilsson — APPROVED** [review]

**Pull request merged**